### PR TITLE
feat: initial instance and collection(s) implementation

### PR DESCRIPTION
**Pull request by MatthewLemmond**
### Description
Create an instance and optionally create collections based on inputs provided, TODO add documentation and outputs, setup testing, add other app config resources as needed.

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [ ] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### Merge actions for mergers

- When merging, use a relevant [conventional commit](https://www.conventionalcommits.org/) message that is based on the PR contents and any release notes provided by the PR author. The commit message determines whether a new version of the module is needed, and if so, which semver increment to use (major, minor, or patch).
- Merge by using "Squash and merge".

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,81 +23,97 @@ https://terraform-ibm-modules.github.io/documentation/#/implementation-guideline
...
-
-Use real values instead of "var.<var_name>" or other placeholder values
-unless real values don't help users know what to change.
--->
-
 ```hcl
-
+module "app_config" {
+  source                       = "terraform-ibm-modules/app-configuration/ibm"
+  version                      = "latest" # Replace "latest" with a release version to lock into a specific release
```
Comment (ocofaigh) [inline README.md:40]: Can you replace "latest" with "X.X.X" as we got feedback that this was confusing and someone thought they could just use the string "latest" to pull latest

File: examples/complete/main.tf [hunk 2]
```diff
@@ -1,3 +1,35 @@
...
 # Complete example
 ##############################################################################
+
+########################################################################################################################
+# Resource group
+########################################################################################################################
+
+module "resource_group" {
+  source  = "terraform-ibm-modules/resource-group/ibm"
+  version = "1.0.6"
```
Comment (ocofaigh) [inline examples/complete/main.tf:11]: bump to 1.1.0

File: tests/go.mod [hunk 3]
```diff
@@ -1,4 +1,4 @@
-module github.com/terraform-ibm-modules/terraform-ibm-module-template
+module github.com/terraform-ibm-modules/terraform-ibm-app-configuration
 
 go 1.20
```
Comment (ocofaigh) [inline tests/go.mod:3]: can you update this to go 1.21 and re-run `make go-mod-tidy` locally? (Ensure your locally on 1.21 before doing this)

Comment (ocofaigh) [reply tests/go.mod:3]: Also update the test wrapper version to the current latest as it has some important fixes

File: variables.tf [hunk 4]
```diff
@@ -1,9 +1,67 @@
...
-#}
+variable "app_config_name" {
+  type        = string
+  description = "Name for the App Configuration service instance"
+}
+
+variable "app_config_plan" {
+  type        = string
+  description = "Plan for the App Configuration service instance"
+  default     = "standardv2"
```
Comment (ocofaigh) [inline variables.tf:33]: perhaps default it to lite

File: variables.tf [hunk 5]
```diff
@@ -1,9 +1,67 @@
...
+
+  validation {
+    condition     = contains(["lite", "standardv2", "enterprise"], var.app_config_plan)
+    error_message = "Value for plan must be one of the following: \"lite\", \"standard\", or \"enterprise\"."
```
Comment (ocofaigh) [inline variables.tf:37]: Is it `standard` or `standardv2`? The condition says `standardv2` bit the error message has `standard`

File: variables.tf [hunk 6]
```diff
@@ -1,9 +1,67 @@
...
+  }
+}
+
+variable "app_config_service_endpoints" {
+  type        = string
+  description = "Service Endpoints for the App Configuration service instance"
+  default     = "public"
+
+  validation {
+    condition     = contains(["public", "private", "public-and-private"], var.app_config_service_endpoints)
```
Comment (ocofaigh) [inline variables.tf:47]: Does it support private only? Did you try it? I see this in the UI:
image

File: variables.tf [hunk 7]
```diff
@@ -1,9 +1,67 @@
...
+  default     = []
+}
+
+variable "app_config_collections" {
+  description = "A list of collections to be added to the App Configuration instance"
+  type = list(object({
+    name          = string
+    collection_id = string
+    description   = optional(string, "")
+    tags          = optional(string, "")
```
Comment (ocofaigh) [inline variables.tf:64]: If description and tags are optional, then why do we have a default value of empty string?

**Review by ocofaigh — CHANGES_REQUESTED** [review]
Left some comments

File: variables.tf (hunk 5 above)
Comment (MatthewLemmond) [reply variables.tf:37]: "standardv2" is correct, updated error message to reflect. I also had considered allowing "standard" and then appending the "v2" if that plan was chosen but decided against it as it seems better to require the inputs match what the provider expects.

File: variables.tf (hunk 6 above)
Comment (MatthewLemmond) [reply variables.tf:47]: I misread docs, based on https://cloud.ibm.com/docs/app-configuration?topic=app-configuration-ac-create-an-instance this only support both and even then it's only for enterprise plans, fixing this var and adding validation in the module to check it's enterprise if both are enabled.

File: variables.tf (hunk 7 above)
Comment (MatthewLemmond) [reply variables.tf:64]: changed to null

File: README.md [hunk 8]
```diff
@@ -36,46 +29,47 @@ https://terraform-ibm-modules.github.io/documentation/#/implementation-guideline
...
 ### Usage
 
-<!--
-Add an example of the use of the module in the following code block.
-
-Use real values instead of "var.<var_name>" or other placeholder values
-unless real values don't help users know what to change.
--->
-
 ```hcl
```
Comment (ocofaigh) [inline README.md:32]: You seem to be missing the module title heading after Overview but before Usage. See module template -> https://github.com/terraform-ibm-modules/terraform-ibm-module-template/blob/619aa36ccb4860c97ced96735b6c33df144e22d3/README.md?plain=1#L40

File: tests/README.md [hunk 9]
```diff
@@ -3,3 +3,4 @@
 For information about how to create and run tests, see [Validation tests](https://terraform-ibm-modules.github.io/documentation/#/tests) in the project documentation.
 
 <!-- Add any more steps that are specific to testing this module and that are not in the docs. -->
+The App Configuration service is only available in particular regions, in order to ensure tests are run in a valid region the environment variable `FORCE_TEST_REGION` must be set to a valid region.
```
Comment (ocofaigh) [inline tests/README.md:6]: Why would someone need to use this? The test itself already only allows particular regions

File: variables.tf [hunk 10]
```diff
@@ -1,9 +1,67 @@
...
+# Common variables
+##############################################################################
+
+variable "resource_group_id" {
+  type        = string
+  description = "The resource group ID where resources will be provisioned."
+}
+
+variable "region" {
+  description = "The region to provision the resources."
```
Comment (ocofaigh) [inline variables.tf:11]: Perhaps mention the supported regions in the description.

File: variables.tf [hunk 11]
```diff
@@ -1,9 +1,67 @@
...
-#  default     = "default_value"
-#}
+variable "app_config_name" {
+  type        = string
+  description = "Name for the App Configuration service instance"
+}
+
+variable "app_config_plan" {
+  type        = string
+  description = "Plan for the App Configuration service instance"
```
Comment (ocofaigh) [inline variables.tf:32]: mention the supported values in the descrption

File: variables.tf [hunk 12]
```diff
@@ -1,9 +1,67 @@
...
+
+  validation {
+    condition     = contains(["lite", "standardv2", "enterprise"], var.app_config_plan)
+    error_message = "Value for plan must be one of the following: \"lite\", \"standardv2\", or \"enterprise\"."
+  }
+}
+
+variable "app_config_service_endpoints" {
+  type        = string
+  description = "Service Endpoints for the App Configuration service instance"
```
Comment (ocofaigh) [inline variables.tf:43]: mention the supported values in the descrption

File: variables.tf [hunk 13]
```diff
@@ -1,9 +1,67 @@
...
+  default     = "public"
```
Comment (ocofaigh) [inline variables.tf:44]: I would default it to `public-and-private`

**Review by ocofaigh — CHANGES_REQUESTED** [review]
Final few comments

File: tests/README.md (hunk 9 above)
Comment (MatthewLemmond) [reply tests/README.md:6]: This was added prior to restricting the test to particular regions, removed.

**Pull request merged**